### PR TITLE
[Merged by Bors] - TY-2726 key phrase selection tests

### PR DESCRIPTION
**Pull request by janpetschexain**
**References**

- [TY-2726]
- #299, #302

**Summary**

- simplify test setup
- add tests for the remaining uncovered key phrase selection subfunctionality
- make `reduce_without_diag()` more robust in edge cases
- simplify `reduce_without_diag()` and `argmax()` function signatures
- generalize `is_selected()` and `select()` function signatures

[TY-2726]: https://xainag.atlassian.net/browse/TY-2726?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: discovery_engine_core/ai/xayn-ai/src/coi/key_phrase.rs [hunk 1]
```diff
@@ -596,18 +899,73 @@ mod tests {
...
+        let mut key_phrases = KeyPhrases::new([
+            (cois[0].id, "key", [1., 1., 1.], ("AA", "aa")),
+            (cois[0].id, "phrase", [2., 1., 1.], ("AA", "aa")),
+            (cois[0].id, "words", [3., 1., 1.], ("AA", "aa")),
+            (cois[0].id, "and", [4., 1., 1.], ("AA", "aa")),
+            (cois[1].id, "more", [1., 6., 1.], ("AA", "aa")),
+            (cois[1].id, "stuff", [1., 5., 1.], ("AA", "aa")),
+        ]);
+        let config = Config::default();
+
```
Comment (rustonaut) [inline discovery_engine_core/ai/xayn-ai/src/coi/key_phrase.rs:942]: :wrench:? This test depends on the default config values, maybe we should assert them to be equal to the values we expect to have easier to understand failures when we change the default values;

**Review by rustonaut — CHANGES_REQUESTED** [review]
Mostly lgtm, but I found one thing in the tested code which probably should be changed and get an additional test.

Also two other minor things which don't need to be changed, but it probably would be better to do so.

File: discovery_engine_core/ai/xayn-ai/src/coi/key_phrase.rs [hunk 2]
```diff
@@ -252,18 +253,14 @@ where
                 .filter_map(|(j, element)| (i != j).then(|| prepare(*element)))
                 .reduce(|x, y| reduce(x, y))
                 .map(|reduced| finalize(reduced, i < lane.len()))
```
[the remark this answers is not in the document]
Comment (janpetschexain) [reply discovery_engine_core/ai/xayn-ai/src/coi/key_phrase.rs:255]: this is what i wanted to say as well, i'll try to find a better phrasing of the documentation.

**Review by rustonaut — APPROVED** [review]
lgtm

**Comment by janpetschexain** [comment]
bors merge
